File: java/demos/ps/space/Space.java
package demos.ps.space;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import static core.Script.*;
import demos.ps.oo.City;
import demos.ps.oo.PSGame;
import demos.ps.oo.PSGame.Flags;
import demos.ps.oo.PSGame.GameType;
import demos.ps.oo.PSGame.Planet;
import demos.ps.oo.PSLibImage.PS1Image;
import demos.ps.oo.PSLibMusic.PS1Music;
import demos.ps.oo.PSMenu.EntityType;
import demos.ps.oo.PSMenu.Scene;
import demos.ps.oo.PSMenu;
import demos.ps.oo.PSMenu.SpecialEntity;
import core.Script;
import domain.Entity;
import domain.VImage;

public class Space {

	private static final Logger log = LogManager.getLogger(Space.class);

	public static void startmap() {

		// INTRO
		if(PSGame.getFromCity() == null) {
			intro();
			return;
		}

		cameratracking = 1;
		
		if(PSGame.getFromCity() == City.CAMINEET || PSGame.getFromCity() == City.GOTHIC) {
			int ent = entityspawn(5, 96, "space/Palma.chr");
			entities.get(ent).setFace (Entity.NORTH);
		}
		if(PSGame.getFromCity() == City.PASEO || PSGame.getFromCity() == City.UZO) {
			int ent = entityspawn(5, 96, "space/Mota.chr");
			entities.get(ent).setFace (Entity.NORTH);
		}
		if(PSGame.getFromCity() == City.SKURE) {
			int ent = entityspawn(5, 96, "space/Dezo.chr");
			entities.get(ent).setFace (Entity.NORTH);
		}
		
		if(PSGame.getToCity() == City.CAMINEET || PSGame.getToCity() == City.GOTHIC) {
			int ent = entityspawn(5, 0, "space/Palma.chr");
			entities.get(ent).setFace (Entity.SOUTH);
		}
		if(PSGame.getToCity() == City.PASEO || PSGame.getToCity() == City.UZO) {
			int ent = entityspawn(5, 0, "space/Mota.chr");
			entities.get(ent).setFace (Entity.SOUTH);
		}
		if(PSGame.getToCity() == City.SKURE) {
			int ent = entityspawn(5, 0, "space/Dezo.chr");
			entities.get(ent).setFace (Entity.SOUTH);
		}		

		PSGame.spaceshipRoutineEnd();
	}
	
	
	public static void intro() {
		int ent = entityspawn(5, 96, "space/Palma.chr");
		entities.get(ent).setFace (Entity.NORTH);

		PSGame.playMusic(PS1Music.INTRO);
		
		cameratracking = 0;
		
		Script.xwin = 0;
		Script.ywin = 1200;
		int count = 0;
		while(count++ < 250) {
			ywin++;
			screen.render();
			showpage();
		}
		
		if(PSGame.getGameType() == GameType.PS_ORIGINAL) {
			alisIntro();
		}
		else if(PSGame.getGameType() == GameType.PS_START_AS_ODIN) {
			odinIntro();
		}
		else if(PSGame.getGameType() == GameType.PS_START_AS_NOAH) {
			noahIntro();
		}
		
	}


	private static void alisIntro() {
		PSMenu.startScene(Scene.CITY, SpecialEntity.NONE);
		PSMenu.instance.push(PSMenu.instance.createOneLabelBox(15, 20, PSGame.getString("Intro_Time"), true));
		PSMenu.instance.waitDelay(20);
		PSMenu.instance.push(PSMenu.instance.createOneLabelBox(15, 50, PSGame.getString("Intro_Place"), true));
		PSMenu.instance.waitDelay(150);
		PSMenu.instance.pop();
		PSMenu.instance.pop();
		//PSMenu.Stext( + "\n" + PSGame.getString("Intro_Place"));
		PSMenu.setMapOff();

		screen.fadeOut(25, false);
		PSMenu.instance.back = new VImage(screen.width, screen.height); ;
		//screen.paintBlack();
		
		PSMenu.cinematicText(PSGame.getImage(PS1Image.CINE_NERO1), 
				new String[]{PSGame.getString("Cinematic_Intro_1"), PSGame.getString("Cinematic_Intro_2"), PSGame.getString("Cinematic_Intro_3")});

		PSMenu.cinematicText(PSGame.getImage(PS1Image.CINE_NERO2), 
				new String[]{PSGame.getString("Cinematic_Intro_4"), PSGame.getString("Cinematic_Intro_5")});

		PSMenu.cinematicText(PSGame.getImage(PS1Image.CINE_PROMISE), 
				new String[]{PSGame.getString("Cinematic_Intro_6")});

		PSMenu.endScene();
		


		PSGame.gameData.current_planet = Planet.PALMA;
		PSGame.mapswitch(City.CAMINEET, 29, 9); // Alis's house
	}
	

	private static void odinIntro() {
		screen.fadeOut(25, false);
		PSMenu.setMapOff();
		PSMenu.instance.back = new VImage(screen.width, screen.height); ;
		//screen.paintBlack();
		
		PSMenu.cinematicText(PSGame.getImage(PS1Image.CINE_INTRO1), 
				new String[]{PSGame.getString("Cinematic_Intro_Odin_1")});

		PSMenu.cinematicText(PSGame.getImage(PS1Image.CINE_INTRO2), 
				new String[]{PSGame.getString("Cinematic_Intro_Odin_2"), PSGame.getString("Cinematic_Intro_Odin_3")});
		
		PSMenu.cinematicText(PSGame.getImage(PS1Image.CINE_ODIN2), 
				new String[]{PSGame.getString("Cinematic_Intro_Odin_4")});

		PSMenu.endScene();		
		
		PSGame.setFlag(Flags.GOT_ODIN);
		PSGame.gameData.current_planet = Planet.PALMA;
		PSGame.mapswitch(City.SCION, 9, 16);
	}

	private static void noahIntro() {
		screen.fadeOut(25, false);
		PSMenu.setMapOff();
		PSMenu.instance.back = new VImage(screen.width, screen.height); ;
		//screen.paintBlack();
		
		//PSMenu.cinematicText(PSGame.getImage(PS1Image.CINE_INTRO1), 
		//		new String[]{PSGame.getString("Cinematic_Intro_Odin_1")});

		PSMenu.cinematicText(PSGame.getImage(PS1Image.CINE_INTRO2), 
				new String[]{PSGame.getString("Cinematic_Intro_Odin_2"), PSGame.getString("Cinematic_Intro_Odin_3")});
		
		//PSMenu.cinematicText(PSGame.getImage(PS1Image.CINE_ODIN2), 
			//	new String[]{PSGame.getString("Cinematic_Intro_Odin_4")});

		PSMenu.endScene();		
		
		PSGame.gameData.current_planet = Planet.MOTAVIA;
		PSGame.mapswitch(Planet.MOTAVIA, 82, 34);
		
	}
	
	
	
}

File: src/demos/ps/space/Space.ts
/**
 * Space - interplanetary travel map script
 * TypeScript port of Space.java. Reached from the Hapsby spaceship launch
 * (spaceshipRoutineAnimation); also plays the new-game intro when no
 * departure city is set.
 */

import { PSGame } from '../PSGame';
import { GameType, Flags } from '../game/GameData';
import { Planet, City } from '../game/City';
import { PS1Image } from '../game/PSLibImage';
import { PS1Music } from '../game/PSLibMusic';
import { PSSceneType, SpecialEntity, PSMenu } from '../PSMenu';
import { MainEngine } from '../../../core/MainEngine';
import { ScriptEngine } from '../../../core/ScriptEngine';
import { EntityDirection } from '../../../domain/Entity';

const SPACE_BASE = 'src/demos/ps/space';

export class Space {

  public static async startmap(): Promise<void> {
    MainEngine.setScriptContext(Space);

    // INTRO
    if (PSGame.getFromCity() === null) {
      await Space.intro();
      return;
    }

    MainEngine.setCameraTracking(1);

    const currentScene = PSGame.getCurrentScene();
    if (!currentScene) return;

    if (PSGame.getFromCity() === City.CAMINEET || PSGame.getFromCity() === City.GOTHIC) {
      const ent = await MainEngine.entityspawn(currentScene, 5, 96, 'Palma.anim.json', SPACE_BASE);
      MainEngine.getEntity(ent)?.setFace(EntityDirection.NORTH);
    }
    if (PSGame.getFromCity() === City.PASEO || PSGame.getFromCity() === City.UZO) {
      const ent = await MainEngine.entityspawn(currentScene, 5, 96, 'Mota.anim.json', SPACE_BASE);
      MainEngine.getEntity(ent)?.setFace(EntityDirection.NORTH);
    }
    if (PSGame.getFromCity() === City.SKURE) {
      const ent = await MainEngine.entityspawn(currentScene, 5, 96, 'Dezo.anim.json', SPACE_BASE);
      MainEngine.getEntity(ent)?.setFace(EntityDirection.NORTH);
    }

    if (PSGame.getToCity() === City.CAMINEET || PSGame.getToCity() === City.GOTHIC) {
      const ent = await MainEngine.entityspawn(currentScene, 5, 0, 'Palma.anim.json', SPACE_BASE);
      MainEngine.getEntity(ent)?.setFace(EntityDirection.SOUTH);
    }
    if (PSGame.getToCity() === City.PASEO || PSGame.getToCity() === City.UZO) {
      const ent = await MainEngine.entityspawn(currentScene, 5, 0, 'Mota.anim.json', SPACE_BASE);
      MainEngine.getEntity(ent)?.setFace(EntityDirection.SOUTH);
    }
    if (PSGame.getToCity() === City.SKURE) {
      const ent = await MainEngine.entityspawn(currentScene, 5, 0, 'Dezo.anim.json', SPACE_BASE);
      MainEngine.getEntity(ent)?.setFace(EntityDirection.SOUTH);
    }

    await PSGame.spaceshipRoutineEnd();
  }

  public static async intro(): Promise<void> {
    const currentScene = PSGame.getCurrentScene();
    if (!currentScene) return;

    const ent = await MainEngine.entityspawn(currentScene, 5, 96, 'Palma.anim.json', SPACE_BASE);
    MainEngine.getEntity(ent)?.setFace(EntityDirection.NORTH);

    await PSGame.playMusic(PS1Music.INTRO);

    MainEngine.setCameraTracking(0);

    // Java: ywin scrolls from 1200 downward for 250 frames (b2 skips ahead).
    // The Java renderer clamps ywin to the map edge, so the scroll saturates
    // with the planet at the bottom of the screen; Phaser's setScroll doesn't
    // clamp, so do it here or the camera overshoots and centers the planet.
    const camera = currentScene.cameras.main;
    const inputManager = (currentScene as any)?.inputManager;
    const map = MainEngine.getCurrentMap();
    const maxScrollY = map ? map.getHeight() * map.getTileHeight() - 240 : 1360;
    camera.setScroll(0, 1200);
    await ScriptEngine.fadein(1, true);
    for (let count = 0; count < 250; count++) {
      if (inputManager?.b2) {
        inputManager.unpress(6); // b2 — held press must not leak into the next part
        camera.setScroll(0, Math.min(1200 + 249, maxScrollY));
        break;
      }
      camera.setScroll(0, Math.min(1200 + count, maxScrollY));
      await PSGame.waitFrames(1);
    }

    if (PSGame.getGameType() === GameType.PS_ORIGINAL) {
      await Space.alisIntro();
    } else if (PSGame.getGameType() === GameType.PS_START_AS_ODIN) {
      await Space.odinIntro();
    } else if (PSGame.getGameType() === GameType.PS_START_AS_NOAH) {
      await Space.noahIntro();
    }
  }

  private static async alisIntro(): Promise<void> {
    await PSMenu.startScene(PSSceneType.CITY, SpecialEntity.NONE);
    PSMenu.instance.push(PSMenu.instance.createOneLabelBox(15, 20, PSGame.getString("Intro_Time"), true));
    await PSMenu.instance.waitDelay(20, true);
    PSMenu.instance.push(PSMenu.instance.createOneLabelBox(15, 50, PSGame.getString("Intro_Place"), true));
    await PSMenu.instance.waitDelay(150, true);
    PSMenu.instance.pop();
    PSMenu.instance.pop();
    PSMenu.setMapOff();

    // Java: PSMenu.instance.back = new VImage(...) — black backdrop. A camera
    // fade would sit on TOP of the portraits (all we'd see is black), so swap
    // in the backdrop and clear the fade with an instant fadein (see Paseo).
    await ScriptEngine.fadeout(25, false);
    PSMenu.instance.setBlackBackground();
    await ScriptEngine.fadein(1, false);

    await PSMenu.cinematicText(await PSGame.getVImage(PS1Image.CINE_NERO1),
      [PSGame.getString("Cinematic_Intro_1"), PSGame.getString("Cinematic_Intro_2"), PSGame.getString("Cinematic_Intro_3")], true);

    await PSMenu.cinematicText(await PSGame.getVImage(PS1Image.CINE_NERO2),
      [PSGame.getString("Cinematic_Intro_4"), PSGame.getString("Cinematic_Intro_5")], true);

    await PSMenu.cinematicText(await PSGame.getVImage(PS1Image.CINE_PROMISE),
      [PSGame.getString("Cinematic_Intro_6")], true);

    // Java's endScene(FADE_HOUSE) ran after a deferred mapswitch, so its
    // fade-in landed on the new map; here endScene would flash the Space map
    // before Camineet loads — close to black and let the mapswitch reveal.
    await PSMenu.endSceneToBlack();

    PSGame.gameData.current_planet = Planet.PALMA;
    await PSGame.mapswitchToCity(City.CAMINEET, 29, 9); // Alis's house
  }

  private static async odinIntro(): Promise<void> {
    await ScriptEngine.fadeout(25, false);
    PSMenu.setMapOff();
    // Java: black-VImage backdrop; camera fade must be cleared or it covers
    // the portraits (see alisIntro)
    PSMenu.instance.setBlackBackground();
    await ScriptEngine.fadein(1, false);

    await PSMenu.cinematicText(await PSGame.getVImage(PS1Image.CINE_INTRO1),
      [PSGame.getString("Cinematic_Intro_Odin_1")], true);

    await PSMenu.cinematicText(await PSGame.getVImage(PS1Image.CINE_INTRO2),
      [PSGame.getString("Cinematic_Intro_Odin_2"), PSGame.getString("Cinematic_Intro_Odin_3")], true);

    await PSMenu.cinematicText(await PSGame.getVImage(PS1Image.CINE_ODIN2),
      [PSGame.getString("Cinematic_Intro_Odin_4")], true);

    // Close to black — endScene would flash the Space map (see alisIntro)
    await PSMenu.endSceneToBlack();

    PSGame.setFlag(Flags.GOT_ODIN);
    PSGame.gameData.current_planet = Planet.PALMA;
    await PSGame.mapswitchToCity(City.SCION, 9, 16);
  }

  private static async noahIntro(): Promise<void> {
    await ScriptEngine.fadeout(25, false);
    PSMenu.setMapOff();
    // Java: black-VImage backdrop; camera fade must be cleared or it covers
    // the portraits (see alisIntro)
    PSMenu.instance.setBlackBackground();
    await ScriptEngine.fadein(1, false);

    await PSMenu.cinematicText(await PSGame.getVImage(PS1Image.CINE_INTRO2),
      [PSGame.getString("Cinematic_Intro_Odin_2"), PSGame.getString("Cinematic_Intro_Odin_3")], true);

    // Close to black — endScene would flash the Space map (see alisIntro)
    await PSMenu.endSceneToBlack();

    PSGame.gameData.current_planet = Planet.MOTAVIA;
    await PSGame.mapswitchToPlanet(Planet.MOTAVIA, 82, 34);
  }
}
